src/cc/libbpf.c

```diff
@@ -1395,3 +1394,59 @@ int bpf_close_perf_event_fd(int fd) {
   }
   return error;
 }
+
+static void check_on_each_cpu(int cpu, struct perf_event_attr *attr, int fd, int pid) {
+  cpu_set_t set;
+  __u64 value;
+  
+  /* Move to target CPU */
+  CPU_ZERO(&set);
+  CPU_SET(cpu, &set);
+  assert(sched_setaffinity(0, sizeof(set), &set) == 0);
+  
+  enum bpf_prog_type prog_type = BPF_PROG_TYPE_PERF_EVENT;
+	int efd = syscall(__NR_perf_event_open, attr, pid, cpu, -1, PERF_FLAG_FD_CLOEXEC);
+	if (efd < 0) {
+    printf("event fd %d err %s\n", efd, strerror(errno));
+    return;
+  }
+
+  int err;
+  err = ioctl(efd, PERF_EVENT_IOC_RESET,0);
+  if (err < 0) {
+    printf("ioctl PERF_EVENT_IOC_RESET failed err number:%d %s\n", errno, strerror(errno));
+    return;
+  }
+
+  err = ioctl(efd, PERF_EVENT_IOC_ENABLE, 0);
+  if (err < 0) {
+    printf("ioctl PERF_EVENT_IOC_ENABLE failed err number:%d %s\n", errno, strerror(errno));
+    return;
+  }
+
+  err = ioctl(efd, PERF_EVENT_IOC_SET_BPF, fd);
+  if (err < 0) {
+    printf("ioctl PERF_EVENT_IOC_SET_BPF failed err number:%d %s\n", errno, strerror(errno));
+    return;
+  }
+}
+
+void bpf_attach_breakpoint(uint64_t symbol_addr, int pid, int progfd, int bp_type) {
+	struct perf_event_attr attr = {};
+  
+  memset(&attr, 0, sizeof(attr));
+  attr.size = sizeof(attr);
+  attr.type = PERF_TYPE_BREAKPOINT;
+  attr.bp_len = HW_BREAKPOINT_LEN_4;
+  attr.bp_addr = symbol_addr;
+	attr.bp_type = (__u32)bp_type;
+  attr.sample_period = 1;
+  attr.precise_ip = 2; // request synchronous delivery
+  attr.wakeup_events = 1;
+
+	int i, nr_cpus = sysconf(_SC_NPROCESSORS_CONF);
+
+  for (i=0; i<nr_cpus; i++) {
+    check_on_each_cpu(i, &attr, progfd, pid);
+  }
```

Comment:
The whole loop above is not needed. You can take a look at @danobi bpftrace prototyping. https://github.com/iovisor/bpftrace/pull/790/. Just call perf_event_open once with pid, cpu = -1, it will have effect on all cpus.

Reply:
Actually as per my experiments, I figured out that breakpoint needs to be applied at each CPU. I tried with cpu=-1 earlier, it worked only few times when the cpu on which breakpoint was applied and the cpu on which the address was accessed are the same.

Reply:
Oh. Thanks! I did not do any experiments and it seems https://github.com/iovisor/bpftrace/pull/790 need change as well. Maybe we can refactor a little bit to use existing API instead of `check_on_each_cpu`.

Reply:
Looked at the kernel tools code,
https://github.com/torvalds/linux/blob/master/tools/perf/util/parse-events.c#L887-L912
https://github.com/torvalds/linux/blob/master/tools/perf/util/parse-events.c#L316-L347
It does try to find the list of cpus which may apply to the breakpoint.

Also, all the self test examples related breakpoint is for the current process on all cpus:
  pid = 0, cpu = -1

But it does look like kernel accepts pid > 0, cpu = -1 for breakpoint. I did not see error returns. Not sure how kernel interprets this. Need a little more study.

Reply:
I'm actually checking for pid_max, it doesn't work unless we apply breakpoint on all CPUs.